### DISCUSSION TITLE
How to test new code as we crate smash

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1454] [General]
Any ideas on how to test new code when we depend on released crates? E.g. when we patch `hashes` if changes are required to `bitcoin` and/or secp when do we test those changes? I've only come up so far with unsatisfying solutions

- patches to crates lower down the dep tree require a draft PR that tests them (this implies doing so for secp _as well as_ bitcoin)
- local testing only with a comment saying "I built this and the changeset required in secp/bitcoin was ..."
- soon as we release crates all crates in the dep tree above them start using github + commit hash for dependencies and we have to keep updating this as PRs merge

I'm guessing that with the Rust community favouring many small crates we are not the first to hit this, am I missing something obvious?

**Comment by Kixunil** [comment]
Patch worked fine for me - I just didn't update `secp256k1`, which should be the only offender since it's "in the middle".

> am I missing something obvious?

I know of no project that has this kind of "dependency hole". Maybe we could have a forked secp256k1 repo where we would all have the access so we could patch deps and then @apoelstra merges after review.
